// @percy/storybook/dist/commands/storybook/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Storybook = void 0;

var _fs = require("fs");

var _logger = _interopRequireDefault(require("@percy/logger"));

var _command = _interopRequireDefault(require("../../command"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Storybook extends _command.default {
  constructor(...args) {
    super(...args);
    this.log = (0, _logger.default)('cli:storybook');
  }

  // Serves a static directory and resolves when listening.
  async startStorybook() {
    let done = async url => {
      let logTimeout = setTimeout(this.log.warn, 3000, 'Waiting on a response from Storybook...');
      await this.setStorybookUrl(url);
      clearTimeout(logTimeout);
    };

    let dir = this.args.url_or_build_dir;
    if (/^https?:\/\//.test(dir)) return done(dir);
    if (!(0, _fs.existsSync)(dir)) return this.error(`Not found: ${dir}`);
    this.log.debug(`Serving Storybook build: ${dir}`);

    let http = require('http');

    let serve = require('serve-handler');

    return new Promise((resolve, reject) => {
      let server = http.createServer((req, res) => {
        serve(req, res, {
          public: dir,
          cleanUrls: false
        });
      }).on('error', reject).listen(() => {
        done(`http://localhost:${server.address().port}`).then(() => resolve(server), reject);
      });
    });
  }

}

exports.Storybook = Storybook;
Storybook.description = 'Snapshot static or hosted Storybook stories';
Storybook.args = [{
  name: 'url_or_build_dir',
  description: 'storybook url or build output directory',
  required: true
}];
Storybook.flags = _command.default.flags;
Storybook.examples = ['$ percy storybook ./build', '$ percy storybook http://localhost:9000/'];
