// @percy/storybook/dist/commands/storybook/start.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Start = void 0;

var _cliCommand = require("@percy/cli-command");

var _logger = _interopRequireDefault(require("@percy/logger"));

var _command = _interopRequireDefault(require("../../command"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Start extends _command.default {
  constructor(...args) {
    super(...args);
    this.log = (0, _logger.default)('cli:storybook:start');
  }

  // Starts Storybook in a child process and resolves when ready
  async startStorybook() {
    let {
      host,
      port
    } = this.flags;
    let args = ['--ci', `--host=${host}`, `--port=${port}`];

    let spawn = require('cross-spawn');

    args = args.concat(this.parse(Start).argv);
    this.log.info(`Running "start-storybook ${args.join(' ')}"`);
    return new Promise((resolve, reject) => {
      let proc = spawn('start-storybook', args, {
        stdio: 'inherit'
      });
      proc.on('error', reject);

      if (proc.pid) {
        /* istanbul ignore next: this is a storybook flag we don't need to test */
        let proto = args.includes('--https') ? 'https' : 'http';

        let close = () => proc.kill();

        this.setStorybookUrl(`${proto}://${host}:${port}`).then(() => resolve({
          close
        }), reject);
      }
    });
  }

}

exports.Start = Start;
Start.description = 'Run start-storybook to snapshot stories';
Start.strict = false;
Start.flags = { ..._command.default.flags,
  port: _cliCommand.flags.integer({
    description: 'port to start Storybook',
    default: 9000
  }),
  host: _cliCommand.flags.string({
    description: 'host to start Storybook',
    default: 'localhost'
  })
};
Start.examples = ['$ percy storybook:start', '$ percy storybook:start --port 9000', '$ percy storybook:start --static-dir public'];
