// @percy/storybook/dist/command.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _cliCommand = _interopRequireWildcard(require("@percy/cli-command"));

var _request = _interopRequireDefault(require("@percy/client/dist/request"));

var _utils = require("@percy/client/dist/utils");

var _core = _interopRequireDefault(require("@percy/core"));

var _package = _interopRequireDefault(require("../package.json"));

var _utils2 = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

class StorybookCommand extends _cliCommand.default {
  async run() {
    if (!this.isPercyEnabled()) {
      return this.log.info('Percy is disabled. Skipping snapshots');
    }

    this.percy = new _core.default({ ...this.percyrc(),
      clientInfo: `${_package.default.name}/${_package.default.version}`,
      server: false
    });
    this.storybook = await this.startStorybook();
    let [version, snapshots] = await Promise.all([this.getStorybookVersion(), this.getStorybookSnapshots()]);

    if (!snapshots.length) {
      return this.error('No snapshots found');
    }

    this.percy.client.addEnvironmentInfo(`storybook/${version}`); // patch client to override root resources when JS is enabled

    this.percy.client.sendSnapshot = (buildId, options) => {
      if (options.enableJavaScript && this.preview) {
        Object.assign(options.resources.find(r => r.root), this.preview);
      }

      let {
        sendSnapshot
      } = this.percy.client.constructor.prototype;
      return sendSnapshot.call(this.percy.client, buildId, options);
    };

    await this.percy.start();
    this.percy.snapshot(snapshots);
  } // Called on error, interupt, or after running


  async finally(error) {
    var _this$percy, _this$percy2, _this$storybook;

    await ((_this$percy = this.percy) === null || _this$percy === void 0 ? void 0 : _this$percy.stop(!!error));
    await ((_this$percy2 = this.percy) === null || _this$percy2 === void 0 ? void 0 : _this$percy2.browser.close());
    await ((_this$storybook = this.storybook) === null || _this$storybook === void 0 ? void 0 : _this$storybook.close());
  } // Set the url after ensuring it is reachable while saving the preview response


  async setStorybookUrl(url, log) {
    let previewUrl = new URL('iframe.html', url).href;
    let previewDOM = await (0, _request.default)(previewUrl, {
      headers: (0, _utils2.getAuthHeaders)(this.percy.config.discovery),
      retryNotFound: true,
      interval: 1000,
      retries: 30
    });
    this.url = url;
    this.preview = {
      content: previewDOM,
      sha: (0, _utils.sha256hash)(previewDOM)
    };
  } // Resolves to the discovered Storybook version


  async getStorybookVersion() {
    let aboutUrl = new URL('?path=/settings/about', this.url).href;
    this.log.debug(`Get Storybook version: ${aboutUrl}`);
    /* istanbul ignore next: no instrumenting injected code */

    return this._eval(aboutUrl, async ({
      waitFor
    }, timeout) => {
      // Use an xpath selector to find the appropriate element containing the version number
      let getPath = path => document.evaluate(path, document, null, 9, null).singleNodeValue;

      let headerPath = "//header[starts-with(text(), 'Storybook ')]";
      await waitFor(() => getPath(headerPath), timeout).catch(() => Promise.reject(new Error('Failed to find a Storybook <header> element')));
      return getPath(headerPath).innerText.match(/[-]{0,1}[\d]*[.]{0,1}[\d]+/g, '').join('');
    }).catch(error => {
      this.log.debug('Unable to retrieve Storybook version');
      this.log.debug(error);
      return 'unknown';
    }).then(v => {
      return v;
    });
  } // Resolves to an array of Storybook stories


  async getStorybookSnapshots() {
    let previewUrl = new URL('iframe.html', this.url).href;
    this.log.debug(`Get Storybook stories: ${previewUrl}`);
    /* istanbul ignore next: no instrumenting injected code */

    return this._eval(previewUrl, async ({
      waitFor
    }, timeout) => {
      let serializeRegExp = r => r && [].concat(r).map(r => r.toString());

      await waitFor(() => !!window.__STORYBOOK_CLIENT_API__, timeout).catch(() => Promise.reject(new Error('Storybook object not found on the window. ' + 'Open Storybook and check the console for errors.')));
      return window.__STORYBOOK_CLIENT_API__.raw().map(({
        id,
        kind,
        name,
        parameters
      }) => {
        var _parameters$percy, _parameters$percy2;

        return {
          name: `${kind}: ${name}`,
          ...(parameters === null || parameters === void 0 ? void 0 : parameters.percy),
          include: serializeRegExp(parameters === null || parameters === void 0 ? void 0 : (_parameters$percy = parameters.percy) === null || _parameters$percy === void 0 ? void 0 : _parameters$percy.include),
          exclude: serializeRegExp(parameters === null || parameters === void 0 ? void 0 : (_parameters$percy2 = parameters.percy) === null || _parameters$percy2 === void 0 ? void 0 : _parameters$percy2.exclude),
          id
        };
      });
    }).then(stories => {
      return (0, _utils2.mapStorySnapshots)(stories, {
        config: this.percy.config.storybook,
        url: this.url
      });
    });
  } // Borrows a percy discovery browser page to navigate to a URL and evaluate a function, returning
  // the results and normalizing any thrown errors.


  async _eval(url, fn) {
    let page;

    try {
      // always ensure the browser is launched before borrowing a page
      await (this._launchingBrowser || (this._launchingBrowser = this.percy.browser.launch())); // provide discovery options that may impact how the page loads

      page = await this.percy.browser.page({
        networkIdleTimeout: this.percy.config.discovery.networkIdleTimeout,
        requestHeaders: (0, _utils2.getAuthHeaders)(this.percy.config.discovery),
        userAgent: this.percy.config.discovery.userAgent
      }); // navigate to the URL and evaluate the provided function

      await page.goto(url);
      return await page.eval(fn, 5000);
    } catch (error) {
      /* istanbul ignore next: purposefully not handling real errors */
      if (typeof error !== 'string') throw error; // make eval errors nicer by stripping the stack trace from the message

      throw new Error(error.replace(/^Error:\s(.*?)\n\s{4}at\s.*$/s, '$1'));
    } finally {
      var _page;

      // always clean up and close the page
      await ((_page = page) === null || _page === void 0 ? void 0 : _page.close());
    }
  }

}

exports.default = StorybookCommand;
StorybookCommand.flags = { ..._cliCommand.flags.logging,
  ..._cliCommand.flags.discovery,
  ..._cliCommand.flags.config,
  include: _cliCommand.flags.string({
    char: 'i',
    description: 'pattern matching story names to include in snapshots',
    percyrc: 'storybook.include',
    multiple: true
  }),
  exclude: _cliCommand.flags.string({
    char: 'e',
    description: 'pattern matching story names to exclude from snapshots',
    percyrc: 'storybook.exclude',
    multiple: true
  }),
  'dry-run': _cliCommand.flags.boolean({
    char: 'd',
    description: 'logs snapshots without creating a build'
  })
};
