// @percy/storybook/dist/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.storyParamsSchema = exports.schemas = exports.configSchema = void 0;

var _config = require("@percy/core/dist/config");

const configSchema = {
  storybook: {
    type: 'object',
    additionalProperties: false,
    properties: {
      include: {
        anyOf: [{
          oneOf: [{
            type: 'string'
          }, {
            instanceof: 'RegExp'
          }]
        }, {
          type: 'array',
          items: {
            $ref: '#/properties/include/anyOf/0'
          }
        }]
      },
      exclude: {
        anyOf: [{
          oneOf: [{
            type: 'string'
          }, {
            instanceof: 'RegExp'
          }]
        }, {
          type: 'array',
          items: {
            $ref: '#/properties/exclude/anyOf/0'
          }
        }]
      },
      args: {
        type: 'object',
        normalize: false
      },
      queryParams: {
        type: 'object',
        normalize: false
      },
      waitForSelector: {
        $ref: '/snapshot#/properties/waitForSelector'
      },
      waitForTimeout: {
        $ref: '/snapshot#/properties/waitForTimeout'
      },
      additionalSnapshots: {
        type: 'array',
        items: { ..._config.snapshotSchema.properties.additionalSnapshots.items,
          properties: {
            include: {
              $ref: '#/properties/include'
            },
            exclude: {
              $ref: '#/properties/exclude'
            },
            args: {
              $ref: '#/properties/args'
            },
            queryParams: {
              $ref: '#/properties/queryParams'
            },
            waitForSelector: {
              $ref: '#/properties/waitForSelector'
            },
            waitForTimeout: {
              $ref: '#/properties/waitForTimeout'
            },
            name: {
              $ref: '/snapshot#/properties/additionalSnapshots/items/properties/name'
            },
            prefix: {
              $ref: '/snapshot#/properties/additionalSnapshots/items/properties/prefix'
            },
            suffix: {
              $ref: '/snapshot#/properties/additionalSnapshots/items/properties/suffix'
            }
          }
        }
      }
    }
  }
};
exports.configSchema = configSchema;
const storyParamsSchema = {
  $id: '/storybook',
  type: 'object',
  additionalProperties: false,
  disallow: ['url', 'execute'],
  properties: { ..._config.snapshotSchema.properties,
    ...configSchema.storybook.properties,
    skip: {
      type: 'boolean'
    }
  }
};
exports.storyParamsSchema = storyParamsSchema;
const schemas = [configSchema, storyParamsSchema];
exports.schemas = schemas;
